app/services/rewards/claim.rb

```diff
@@ -0,0 +1,26 @@
+# Refresh all quests
+module Rewards
+  class Claim
+    def initialize(user:, reward:)
+      @user = user
+      # ensure we are using the subclass
+      @reward = reward.becomes(reward.type.constantize)
+    end
+
+    def call
+      # reward out of stock
+      return if @reward.stock <= 0 || @reward.claimed?(@user)
+      # user does not have enough experience
+      return if @user.experience_points.sum(:amount) < @reward.cost
+
+      ActiveRecord::Base.transaction do
```

Comment:
Missing lock with stock check

Reply:
https://api.rubyonrails.org/classes/ActiveRecord/Locking/Pessimistic.html. You can check some ways here to simplify the code.

```suggestion
 account.with_lock do
    # This block is called within a transaction,
    # account is already locked.
    account.balance -= 100
    account.save!
end

```